net/fetcher.go

```diff
@@ -22,6 +30,212 @@ type Fetcher interface {
+				ts, err := gsf.LoadTipsetFromCids(ctx, cidResponseLayer)
+				if err != nil {
+					return nil, err
+				}
+
+				out = append(out, ts)
+				isDone, err := done(ts)
+				if err != nil {
+					return nil, err
+				}
+
+				if isDone {
+					break makeRecursion
+				}
+				baseCid = cidResponseLayer[0]
+			}
+			if recursionDepth < maxRecursionDepth {
+				recursionDepth *= recursionMultiplier
+			}
+		}
+	}
+	return out, nil
+}
+
+// LoadTipsetFromCids loads and sanitizes an already downloaded TipSet
+// from an array of cids
+func (gsf *GraphSyncFetcher) LoadTipsetFromCids(ctx context.Context, cidResponseLayer []cid.Cid) (types.TipSet, error) {
+	blockResponseLayer := make([]blocks.Block, 0, len(cidResponseLayer))
+	for _, cid := range cidResponseLayer {
+		block, err := gsf.bs.Get(cid)
+		if err != nil {
+			return types.UndefTipSet, err
+		}
+		blockResponseLayer = append(blockResponseLayer, block)
+	}
+	validatedBlocks, err := sanitizeBlocks(ctx, blockResponseLayer, gsf.validator)
+	if err != nil {
+		return types.UndefTipSet, err
+	}
+
+	ts, err := types.NewTipSet(validatedBlocks...)
+	if err != nil {
+		return types.UndefTipSet, err
+	}
+
+	return ts, nil
+}
+
+type resultSet struct {
+	responseChan <-chan graphsync.ResponseProgress
+	errChan      <-chan error
+}
+
+// GetSingleLayer requests a single layer of cids as individual bocks, fetching
+// non-recursively
+func (gsf *GraphSyncFetcher) GetSingleLayer(ctx context.Context, cids []cid.Cid, from peer.ID) error {
+	selector, err := gsf.bridge.BuildSelector(func(ssb selector.SelectorSpecBuilder) selector.SelectorSpec {
+		return ssb.Matcher()
+	})
+	if err != nil {
+		return err
+	}
+	resultSets := make([]resultSet, 0, len(cids))
+	for _, c := range cids {
+		responseChan, errChan := gsf.gs.Request(ctx, from, cidlink.Link{Cid: c}, selector)
+		resultSets = append(resultSets, resultSet{responseChan, errChan})
+	}
+	for _, rs := range resultSets {
+		for err := range rs.errChan {
+			return err
+		}
+		for range rs.responseChan {
+		}
+	}
+	return nil
+}
+
+// GetBlocks gets multiple layers of parent blocks starting from a baseCid, up to
+// the given recursion depth parameter
+func (gsf *GraphSyncFetcher) GetBlocks(ctx context.Context, baseCid cid.Cid, from peer.ID, recursionDepth int) ([][]cid.Cid, error) {
+
+	selector, err := gsf.bridge.BuildSelector(func(ssb selector.SelectorSpecBuilder) selector.SelectorSpec {
+		return ssb.ExploreRecursive(recursionDepth, ssb.ExploreFields(func(efsb selector.ExploreFieldsSpecBuilder) {
+			efsb.Insert("parents", ssb.ExploreUnion(
+				ssb.ExploreAll(ssb.Matcher()),
+				ssb.ExploreIndex(0, ssb.ExploreRecursiveEdge()),
+			))
+		}))
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	cidResponseLayers := make([][]cid.Cid, recursionDepth)
+	hasCids := make([]map[cid.Cid]struct{}, recursionDepth)
+
+	responseChan, errChan := gsf.gs.Request(ctx, from, cidlink.Link{Cid: baseCid}, selector)
+	for err := range errChan {
+		return nil, err
+	}
```

Comment:
Does this block until the entire request is complete (and the errChan closed), buffering all results in memory? It seems like there should be a better pattern.